File: src-tauri/src/tools/shell.rs
use crate::db::Db;
use crate::tools::vault::{get_vault_root, get_work_root};
use crate::tools::{
    ToolDefinition, ToolError, ToolExecutionContext, ToolMetadata, ToolRegistry, ToolResultMode,
};
use serde_json::{json, Value};
use std::io::Read;
use std::path::PathBuf;
use std::process::{Command, Stdio};
use std::sync::mpsc;
use std::sync::Arc;
use std::time::Duration;

const DEFAULT_TIMEOUT_MS: u64 = 30_000;
const MAX_OUTPUT_BYTES: usize = 256 * 1024; // 256 KB per stream

pub fn register_shell_tools(registry: &mut ToolRegistry, db: Db) -> Result<(), String> {
    let metadata = ToolMetadata {
        name: "shell.exec".to_string(),
        description: "Execute a shell command and return its output. Use for file operations like \
            splitting files (split), counting lines (wc -l), listing files (ls), etc. \
            Use root=\"vault\" or root=\"work\" to run in the configured vault/work directory. \
            Always requires user approval before running."
            .to_string(),
        args_schema: json!({
            "type": "object",
            "properties": {
                "command": { "type": "string" },
                "args": {
                    "type": "array",
                    "items": { "type": "string" }
                },
                "root": { "type": "string", "enum": ["vault", "work"] },
                "working_dir": { "type": "string" },
                "timeout_ms": { "type": "integer", "minimum": 1 }
            },
            "required": ["command"],
            "additionalProperties": false
        }),
        result_schema: json!({
            "type": "object",
            "properties": {
                "stdout":    { "type": "string" },
                "stderr":    { "type": "string" },
                "exit_code": { "type": "integer" },
                "success":   { "type": "boolean" },
                "truncated": { "type": "boolean" }
            },
            "required": ["stdout", "stderr", "exit_code", "success", "truncated"],
            "additionalProperties": false
        }),
        requires_approval: true,
        result_mode: ToolResultMode::Auto,
    };

    let handler = Arc::new(move |args: Value, _ctx: ToolExecutionContext| {
        let command = require_string_arg(&args, "command")?;
        let cmd_args: Vec<String> = args
            .get("args")
            .and_then(|v| v.as_array())
            .map(|arr| {
                arr.iter()
                    .filter_map(|v| v.as_str().map(|s| s.to_string()))
                    .collect()
            })
            .unwrap_or_default();

        let root = args.get("root").and_then(|v| v.as_str());
        let working_dir_arg = args
            .get("working_dir")
            .and_then(|v| v.as_str())
            .map(|s| s.to_string());

        let resolved_dir: Option<PathBuf> = match root {
            Some("vault") => Some(get_vault_root(&db)?),
            Some("work") => Some(get_work_root(&db)?),
            Some(other) => {
                return Err(ToolError::new(format!(
                    "Invalid root '{other}'; expected 'vault' or 'work'"
                )))
            }
            None => working_dir_arg.map(PathBuf::from),
        };

        if let Some(ref dir) = resolved_dir {
            if !dir.is_dir() {
                return Err(ToolError::new(format!(
                    "working_dir does not exist or is not a directory: {}",
                    dir.display()
                )));
            }
        }

        let timeout_ms = args
            .get("timeout_ms")
            .and_then(|v| v.as_u64())
            .unwrap_or(DEFAULT_TIMEOUT_MS);

        let mut cmd = Command::new(&command);
        cmd.args(&cmd_args);
        cmd.stdout(Stdio::piped());
        cmd.stderr(Stdio::piped());
        if let Some(ref dir) = resolved_dir {
            cmd.current_dir(dir);
        }

        let mut child = cmd
            .spawn()
            .map_err(|err| ToolError::new(format!("Failed to spawn '{command}': {err}")))?;

        let mut stdout_handle = child.stdout.take();
        let mut stderr_handle = child.stderr.take();

        let (tx, rx) = mpsc::channel::<Result<(i32, String, String, bool), String>>();

        std::thread::spawn(move || {
            let mut stdout_bytes = Vec::new();
            let mut stderr_bytes = Vec::new();
            let mut truncated = false;

            if let Some(ref mut out) = stdout_handle {
                let mut buf = vec![0u8; MAX_OUTPUT_BYTES + 1];
                let n = out.read(&mut buf).unwrap_or(0);
                if n > MAX_OUTPUT_BYTES {
                    stdout_bytes.extend_from_slice(&buf[..MAX_OUTPUT_BYTES]);
                    truncated = true;
                } else {
                    stdout_bytes.extend_from_slice(&buf[..n]);
                    // drain the rest so the process can exit
                    let _ = std::io::copy(out, &mut std::io::sink());
                }
            }
            if let Some(ref mut err) = stderr_handle {
                let mut buf = vec![0u8; MAX_OUTPUT_BYTES + 1];
                let n = err.read(&mut buf).unwrap_or(0);
                if n > MAX_OUTPUT_BYTES {
                    stderr_bytes.extend_from_slice(&buf[..MAX_OUTPUT_BYTES]);
                    truncated = true;
                } else {
                    stderr_bytes.extend_from_slice(&buf[..n]);
                    let _ = std::io::copy(err, &mut std::io::sink());
                }
            }

            let exit_code = match child.wait() {
                Ok(status) => status.code().unwrap_or(-1),
                Err(err) => {
                    let _ = tx.send(Err(format!("Failed to wait for process: {err}")));
                    return;
                }
            };

            let stdout = String::from_utf8_lossy(&stdout_bytes).into_owned();
            let stderr = String::from_utf8_lossy(&stderr_bytes).into_owned();
            let _ = tx.send(Ok((exit_code, stdout, stderr, truncated)));
        });

        match rx.recv_timeout(Duration::from_millis(timeout_ms)) {
            Ok(Ok((exit_code, stdout, stderr, truncated))) => Ok(json!({
                "stdout": stdout,
                "stderr": stderr,
                "exit_code": exit_code,
                "success": exit_code == 0,
                "truncated": truncated
            })),
            Ok(Err(err)) => Err(ToolError::new(err)),
            Err(mpsc::RecvTimeoutError::Timeout) => {
                // The thread still owns the child, but we can't kill it from here.
                // Return a timeout result; the child will be dropped when the thread finishes.
                Ok(json!({
                    "stdout": "",
                    "stderr": format!("Command timed out after {timeout_ms}ms"),
                    "exit_code": -1,
                    "success": false,
                    "truncated": false
                }))
            }
            Err(mpsc::RecvTimeoutError::Disconnected) => {
                Err(ToolError::new("Command thread disconnected unexpectedly"))
            }
        }
    });

    registry.register(ToolDefinition {
        metadata,
        handler,
        preview: None,
    })
}

fn require_string_arg(args: &Value, key: &str) -> Result<String, ToolError> {
    args.get(key)
        .and_then(|v| v.as_str())
        .map(|v| v.to_string())
        .ok_or_else(|| ToolError::new(format!("Missing or invalid '{key}'")))
}

File: server/src/tools/shell.ts
import { execFile } from 'child_process'
import { promisify } from 'util'
import type { ToolHandler, ToolResult, ToolContext } from './types.js'

const execFileAsync = promisify(execFile)

const DEFAULT_TIMEOUT_MS = 30_000
const MAX_OUTPUT_BYTES = 256 * 1024 // 256KB

export function registerShellTools(registry: { register: (h: ToolHandler) => void }): void {
  registry.register({
    metadata: {
      name: 'shell.exec',
      description: 'Execute a shell command.',
      parameters: {
        type: 'object',
        properties: {
          command: { type: 'string', description: 'Command' },
          working_dir: { type: 'string', description: 'Working directory' },
          timeout_ms: {
            type: 'integer',
            description: `Default: ${DEFAULT_TIMEOUT_MS}`,
          },
        },
        required: ['command'],
      },
      requires_approval: true,
    },
    async handle(args: Record<string, unknown>, ctx: ToolContext): Promise<ToolResult> {
      const command = args.command as string
      const workingDir = args.working_dir as string | undefined
      const timeoutMs = (args.timeout_ms as number) ?? DEFAULT_TIMEOUT_MS

      const abortController = new AbortController()

      // Link to parent signal
      const onAbort = () => abortController.abort()
      ctx.signal.addEventListener('abort', onAbort, { once: true })

      // Timeout
      const timer = setTimeout(() => abortController.abort(), timeoutMs)

      try {
        const { stdout, stderr } = await execFileAsync('/bin/sh', ['-c', command], {
          cwd: workingDir,
          maxBuffer: MAX_OUTPUT_BYTES,
          signal: abortController.signal,
        })

        return {
          ok: true,
          output: {
            stdout: truncate(stdout, MAX_OUTPUT_BYTES),
            stderr: truncate(stderr, MAX_OUTPUT_BYTES),
            exit_code: 0,
            success: true,
          },
        }
      } catch (err) {
        const error = err as {
          code?: number | string
          stdout?: string
          stderr?: string
          killed?: boolean
          message?: string
        }

        if (error.killed || error.code === 'ABORT_ERR' || ctx.signal.aborted) {
          return {
            ok: false,
            error: ctx.signal.aborted ? 'Command aborted' : 'Command timed out',
            output: {
              stdout: truncate(error.stdout ?? '', MAX_OUTPUT_BYTES),
              stderr: truncate(error.stderr ?? '', MAX_OUTPUT_BYTES),
              exit_code: -1,
              success: false,
            },
          }
        }

        // Non-zero exit code
        const exitCode = typeof error.code === 'number' ? error.code : 1
        return {
          ok: true,
          output: {
            stdout: truncate(error.stdout ?? '', MAX_OUTPUT_BYTES),
            stderr: truncate(error.stderr ?? '', MAX_OUTPUT_BYTES),
            exit_code: exitCode,
            success: false,
          },
        }
      } finally {
        clearTimeout(timer)
        ctx.signal.removeEventListener('abort', onAbort)
      }
    },
    preview(args: Record<string, unknown>): { summary: string; details?: Record<string, unknown> } {
      const cmd = String(args.command)
      const display = cmd.length > 80 ? cmd.slice(0, 77) + '...' : cmd
      return {
        summary: `Execute: ${display}`,
        details: args.working_dir ? { working_dir: args.working_dir } : undefined,
      }
    },
  })
}

function truncate(str: string, maxBytes: number): string {
  if (Buffer.byteLength(str) <= maxBytes) return str
  // Rough truncation — good enough for output capping
  const buf = Buffer.from(str)
  return buf.subarray(0, maxBytes).toString('utf-8') + '\n... [truncated]'
}
